src/rdrand.rs

```diff
@@ -43,25 +36,53 @@ compile_error!(
     "SGX targets require 'rdrand' target feature. Enable by using -C target-feature=+rdrnd."
 );
 
-#[cfg(target_feature = "rdrand")]
+#[cfg(target_env = "sgx")]
 fn is_rdrand_supported() -> bool {
     true
 }
 
 // TODO use is_x86_feature_detected!("rdrand") when that works in core. See:
 // https://github.com/rust-lang-nursery/stdsimd/issues/464
-#[cfg(not(target_feature = "rdrand"))]
+#[cfg(not(target_env = "sgx"))]
 fn is_rdrand_supported() -> bool {
-    use core::arch::x86_64::__cpuid;
     use lazy_static::lazy_static;
-    // SAFETY: All x86_64 CPUs support CPUID leaf 1
-    const FLAG: u32 = 1 << 30;
     lazy_static! {
-        static ref HAS_RDRAND: bool = unsafe { __cpuid(1).ecx & FLAG != 0 };
+        static ref HAS_RDRAND: bool = has_rdrand();
     }
     *HAS_RDRAND
 }
 
+#[cfg(not(target_env = "sgx"))]
+pub fn has_rdrand() -> bool {
+    use core::arch::x86_64::__cpuid;
+    // SAFETY: All x86_64 CPUs support CPUID leaf 0 and 1
+    let leaf_0 = unsafe { __cpuid(0) };
+    let vendor_id: [u8; 12] = unsafe { mem::transmute([leaf_0.ebx, leaf_0.edx, leaf_0.ecx]) };
+    let is_amd = &vendor_id == b"AuthenticAMD";
```

Comment:
You can write this part slightly more efficiently:
```rust
let leaf0 = unsafe { __cpuid(0) };
let amd: &[u8; 12] = b"AuthenticAMD";
let r: [u32; 3] = unsafe { mem::transmute_copy(amd) };
leaf0.ebx == r[0] && leaf0.edx == r[1] && leaf0.ecx == r[2]
```

Reply:
I just got rid of this vendor ID check entirely.